tstest/clock.go

```diff
@@ -4,57 +4,495 @@
+// implements heap.Interface for efficient lookups of the next event.
+type eventManager struct {
+	mu            sync.Mutex
+	now           time.Time
+	heap          []*event
+	reverseLookup map[eventHandler]*event
+}
+
+// Push implements heap.Interface.Push and must only be called by heap funcs
+// with em.mu already held.
+func (em *eventManager) Push(x any) {
+	e, ok := x.(*event)
+	if !ok {
+		panic("incorrect event type")
+	}
+	if e == nil {
+		panic("nil event")
+	}
+
+	mak.Set(&em.reverseLookup, e.eh, e)
+	e.position = len(em.heap)
+	em.heap = append(em.heap, e)
+}
+
+// Pop implements heap.Interface.Pop and must only be called by heap funcs with
+// em.mu already held.
+func (em *eventManager) Pop() any {
+	e := em.heap[len(em.heap)-1]
+	em.heap = em.heap[:len(em.heap)-1]
+	delete(em.reverseLookup, e.eh)
+	return e
+}
+
+// Len implements sort.Interface.Len and must only be called by heap funcs with
+// em.mu already held.
+func (em *eventManager) Len() int {
+	return len(em.heap)
+}
+
+// Less implements sort.Interface.Less and must only be called by heap funcs
+// with em.mu already held.
+func (em *eventManager) Less(i, j int) bool {
+	return em.heap[i].when.Before(em.heap[j].when)
+}
+
+// Swap implements sort.Interface.Swap and must only be called by heap funcs
+// with em.mu already held.
+func (em *eventManager) Swap(i, j int) {
+	em.heap[i], em.heap[j] = em.heap[j], em.heap[i]
+	em.heap[i].position = i
+	em.heap[j].position = j
+}
+
+// Reschedule adds/updates/deletes an event in the heap, whichever
+// operation is applicable (use a zero time to delete).
+func (em *eventManager) Reschedule(eh eventHandler, t time.Time) {
+	em.mu.Lock()
+	defer em.mu.Unlock()
+	e, ok := em.reverseLookup[eh]
+	if !ok {
+		if t.IsZero() {
+			// eh is not scheduled and also not active, so do nothing.
+			return
+		}
+		// eh is not scheduled but is active, so add it.
+		heap.Push(em, &event{
+			when: t,
+			eh:   eh,
+		})
+		return
+	}
+
+	if t.IsZero() {
+		// e is scheduled but not active, so remove it.
+		heap.Remove(em, e.position)
+		return
+	}
+
+	// e is scheduled and active, so update it.
+	e.when = t
+	heap.Fix(em, e.position)
+}
+
+func (em *eventManager) AdvanceTo(tm time.Time) {
+	for len(em.heap) > 0 {
+		if em.heap[0].when.After(tm) {
+			break
+		}
+
+		// Ideally some jitter would be added here but it's difficult to do so
+		// in a deterministic fashion.
+		em.now = em.heap[0].when
+
+		if nextFire := em.heap[0].eh.Fire(em.now); !nextFire.IsZero() {
+			em.heap[0].when = nextFire
+			heap.Fix(em, 0)
+		} else {
+			heap.Pop(em)
+		}
+	}
+
+	em.now = tm
+}
+
+func (em *eventManager) Now() time.Time {
+	em.mu.Lock()
+	defer em.mu.Unlock()
+	return em.now
+}
+
+// Ticker is a time.Ticker lookalike for use in tests that need to control when
+// events fire. Ticker could be made standalone in future but for now is
+// expected to be paired with a Clock and created by Clock.NewTicker.
+type Ticker struct {
```

Comment:
Sort of general comment: do we need a time.AfterFunc equivalent as well? Easy to assemble from a Timer and some glue, just curious.

Reply:
Yeah I ended up needing it anyway to implement the new `FollowsRealTime` option that replicates what `tstester` in corp does.